src/node_quic_session.cc

```diff
@@ -1535,36 +2213,41 @@ void QuicSession::UpdateIdleTimer() {
   idle_->Update(timeout);
 }
 
-// For the server-side, we only care that the client provided
-// certificate is signed by some entity the server trusts.
-// Any additional checks can be performed in usercode on the
-// JavaScript side. QuicClientSession overrides to perform
-// additional checks.
-// The hostname arg is ignored in the default case, but
-// not in the QuicClientSession override.
-int QuicSession::VerifyPeerIdentity(const char* hostname) {
-  return VerifyPeerCertificate(ssl());
-}
-
-void QuicSession::WriteHandshake(
-    ngtcp2_crypto_level level,
-    const uint8_t* data,
-    size_t datalen) {
+void QuicSession::VersionNegotiation(
+      const ngtcp2_pkt_hd* hd,
+      const uint32_t* sv,
+      size_t nsv) {
+  CHECK(!IsServer());
   if (IsFlagSet(QUICSESSION_FLAG_DESTROYED))
     return;
-  Debug(this, "Writing %d bytes of %s handshake data.",
-        datalen,
-        crypto_level_name(level));
-  MallocedBuffer<uint8_t> buffer(datalen);
-  memcpy(buffer.data, data, datalen);
-  // TODO(@jasnell): Need to verify the data ownership here
-  session_stats_.handshake_send_at = uv_hrtime();
-  CHECK_EQ(
-      ngtcp2_conn_submit_crypto_data(
-          Connection(),
-          level,
-          buffer.data, datalen), 0);
-  handshake_[level].Push(std::move(buffer));
+  HandleScope scope(env()->isolate());
+  Local<Context> context = env()->context();
+  Context::Scope context_scope(context);
+
+  Local<Array> versions = Array::New(env()->isolate(), nsv);
+  for (size_t n = 0; n < nsv; n++) {
+    USE(versions->Set(
+        env()->context(), n,
+        Integer::New(env()->isolate(), sv[n])));
+  }
+
+  Local<Array> supportedVersions = Array::New(env()->isolate(), 1);
+  USE(supportedVersions->Set(
+      env()->context(), 0,
+      Integer::New(env()->isolate(), NGTCP2_PROTO_VER)));
```

Comment:
style nit: `snake_case` for local variables

Substantial nit: a) repeated use of `USE()` is not great, as mentioned above, and b) we basically always prefer the 3-argument variant of `Array::New()` that takes an array of `Local<Value>`s for creating arrays this way.